draft-ietf-quic-tls.md

```diff
@@ -1086,13 +1079,25 @@ before the final TLS handshake messages are received.  A client will be unable
 to decrypt 1-RTT packets from the server, whereas a server will be able to
 decrypt 1-RTT packets from the client.
 
-However, a server MUST NOT process data from incoming 1-RTT protected packets
-before verifying either the client Finished message or - in the case that the
-server has chosen to use a pre-shared key - the pre-shared key binder (see
-Section 4.2.11 of {{!TLS13}}).  Verifying these values provides the server with
-an assurance that the ClientHello has not been modified.  Packets protected with
-1-RTT keys MAY be stored and later decrypted and used once the handshake is
-complete.
+In this document we consider the server to have verified the identity of the
+client when either the handshake is complete from the server's perspective,
+or - in the case that the server has chosen to use a pre-shared key - when it
+has validated the pre-shared key binder (see Section 4.2.11 of [TLS13]).
+
+The server has access to 1-RTT keys after receiving the first handshake
+messages from a client, which can happen before it has verified the identity of
+the client.  Therefore, a server MUST NOT process data from incoming 1-RTT
+protected packets before it has verified the client's identity.  In particular,
+it MUST NOT send acknowledgments for 1-RTT packets until then.  Packets
```

Comment:
This MUST NOT send acknowledgements bit is not quite right given the definition of "verified the client's identity" you are using, because that would allow the server to process 1-RTT packets without having seen the client's second flight.  I think that there are two things we need to be pushing on:

1. You need the PSK binder to confirm client identity before you send 0.5 RTT data (that is, any data sent by the server prior to handshake completion).  Before that, you are working with no knowledge of the identity of the client.  Even with a PSK binder, you don't have proof that there isn't an active attack on the connection until the handshake is complete.

2. Handshake completion is necessary for anything.

Here, I would argue that the best option is to wait for handshake completion before using or acknowledging 1-RTT packets.  That's strictly less than what we have allowed for, but it avoids complications.  The point of the text about client identity is to enable use of client authentication - carried over from a resumed connection - in processing of 0-RTT.  It's not a broad license to ignore the rest of the handshake.

Reply:
I agree that waiting until the handshake is complete is safer. I've changed the text to require that.